File: sdk-js/src/verifier.ts
import { TollaraHeaders } from './tollaraHeaders';
import { calculateHmac, constantTimeEquals } from './hmac';
import { grantAccess } from './grantAccess';

export { grantAccess };

export interface UserContext {
  userId: string | null;
  /** v3: service product id; v1/v2 legacy: plan name in {@link plan}. */
  serviceProductId: string | null;
  /** @deprecated v1/v2 only */
  plan: string | null;
  roles: string[];
  /** @deprecated v1 only */
  quotaRemaining: number | null;
  /** v3: uppercase subscription status. */
  subscriptionStatus: string | null;
  /** @deprecated v1/v2 only */
  subscriptionActive: boolean;
  billingModelType: string | null;
  measurementType: string | null;
  unitLabel: string | null;
}

export interface VerifySignatureInput {
  signature: string;
  timestamp: string;
  payload: string | object | null;
  userId: string | null;
  /** v3 */
  serviceProductId?: string | null;
  /** v1/v2 */
  plan?: string | null;
  roles: string[];
  /** v1 only */
  quotaRemaining?: number | string | null;
  /** v3 */
  subscriptionStatus?: string | null;
  /** v1/v2 */
  subscriptionActive?: boolean;
  billingModelType?: string | null;
  measurementType?: string | null;
  unitLabel?: string | null;
  /** `3` = v3; `2` = v2; absent = v1 */
  signingVersion?: string | null;
}

/** User fields that participate in inbound HMAC userContextString. */
export interface SignedUserContext {
  userId: string | null;
  serviceProductId?: string | null;
  plan?: string | null;
  roles: string[];
  quotaRemaining?: number | string | null;
  subscriptionStatus?: string | null;
  subscriptionActive?: boolean;
  billingModelType?: string | null;
  measurementType?: string | null;
  unitLabel?: string | null;
}

export interface InboundHmacRequest {
  signature: string;
  timestamp: string;
  payload: string | object | null;
  signedUserContext: SignedUserContext;
  signingVersion?: string | null;
}

/** Loose header map (e.g. Node/Express lowercases keys). */
export type HeaderBag = Record<string, string | string[] | undefined | null>;

function headerGet(headers: HeaderBag, canonicalName: string): string | null {
  const target = canonicalName.toLowerCase();
  for (const [k, v] of Object.entries(headers)) {
    if (k.toLowerCase() === target) {
      if (v == null) return null;
      return Array.isArray(v) ? (v[0] ?? null) : String(v);
    }
  }
  return null;
}

function parseSubscriptionActive(raw: string | null): boolean {
  if (raw == null || raw === '') return false;
  const t = raw.trim();
  return t === 'true' || t === '1';
}

/**
 * Gateway inbound HMAC suffix v1: userId + plan + rolesCsv + quota + subscriptionActive + billing fields.
 */
export function buildGatewayUserContextString(
  userId: string | null,
  plan: string | null,
  roles: string[],
  quotaRemaining: number | string | null,
  subscriptionActive: boolean,
  billingModelType: string | null,
  measurementType: string | null,
  unitLabel: string | null
): string {
  const u = userId ?? '';
  const p = plan ?? '';
  const r = roles?.length ? roles.join(',') : '';
  const q = quotaRemaining != null && quotaRemaining !== '' ? String(quotaRemaining) : '';
  const sub = subscriptionActive ? 'true' : 'false';
  const b = billingModelType ?? '';
  const m = measurementType ?? '';
  const ul = unitLabel ?? '';
  return u + p + r + q + sub + b + m + ul;
}

/** Gateway HMAC user-context v2: leading `2`, then userId, plan, roles, subscription, billing fields (no quota). */
export function buildGatewayUserContextStringV2(
  userId: string | null,
  plan: string | null,
  roles: string[],
  subscriptionActive: boolean,
  billingModelType: string | null,
  measurementType: string | null,
  unitLabel: string | null
): string {
  const u = userId ?? '';
  const p = plan ?? '';
  const r = roles?.length ? roles.join(',') : '';
  const sub = subscriptionActive ? 'true' : 'false';
  const b = billingModelType ?? '';
  const m = measurementType ?? '';
  const ul = unitLabel ?? '';
  return '2' + u + p + r + sub + b + m + ul;
}

/** Gateway HMAC user-context v3: leading `3`, serviceProductId, subscriptionStatus (see MAIN-SDK-API-SPEC §4). */
export function buildGatewayUserContextStringV3(
  userId: string | null,
  serviceProductId: string | null,
  roles: string[],
  subscriptionStatus: string | null,
  billingModelType: string | null,
  measurementType: string | null,
  unitLabel: string | null
): string {
  const u = userId ?? '';
  const sp = serviceProductId ?? '';
  const r = roles?.length ? roles.join(',') : '';
  const st = subscriptionStatus ?? '';
  const b = billingModelType ?? '';
  const m = measurementType ?? '';
  const ul = unitLabel ?? '';
  return '3' + u + sp + r + st + b + m + ul;
}

function buildUserContextString(input: VerifySignatureInput): string {
  const {
    userId,
    serviceProductId,
    plan,
    roles,
    quotaRemaining,
    subscriptionStatus,
    subscriptionActive,
    billingModelType,
    measurementType,
    unitLabel,
    signingVersion,
  } = input;
  if (signingVersion === '3') {
    return buildGatewayUserContextStringV3(
      userId,
      serviceProductId ?? null,
      roles ?? [],
      subscriptionStatus ?? null,
      billingModelType ?? null,
      measurementType ?? null,
      unitLabel ?? null
    );
  }
  if (signingVersion === '2') {
    return buildGatewayUserContextStringV2(
      userId,
      plan ?? null,
      roles ?? [],
      subscriptionActive ?? false,
      billingModelType ?? null,
      measurementType ?? null,
      unitLabel ?? null
    );
  }
  return buildGatewayUserContextString(
    userId,
    plan ?? null,
    roles ?? [],
    quotaRemaining ?? null,
    subscriptionActive ?? false,
    billingModelType ?? null,
    measurementType ?? null,
    unitLabel ?? null
  );
}

/**
 * Verifies HMAC on an inbound gateway request.
 * Canonical string: payload + timestamp + userContextString (see docs-sdk/MAIN-SDK-API-SPEC.md §4).
 */
export function verifySignature(serviceSecret: string, input: VerifySignatureInput): boolean {
  const { signature, timestamp, payload } = input;
  if (!signature || !timestamp || !serviceSecret) return false;
  try {
    const payloadString =
      payload == null ? '' : typeof payload === 'string' ? payload : JSON.stringify(payload);
    const userContextString = buildUserContextString(input);
    const dataToSign = payloadString + timestamp + userContextString;
    const expectedSignature = calculateHmac(dataToSign, serviceSecret);
    return constantTimeEquals(expectedSignature, signature);
  } catch {
    return false;
  }
}

/**
 * Verifies inbound HMAC using {@link InboundHmacRequest} (preferred typed entry point).
 */
export function verifyInboundHmac(serviceSecret: string, request: InboundHmacRequest): boolean {
  const s = request.signedUserContext;
  return verifySignature(serviceSecret, {
    signature: request.signature,
    timestamp: request.timestamp,
    payload: request.payload,
    userId: s.userId,
    serviceProductId: s.serviceProductId,
    plan: s.plan,
    roles: s.roles ?? [],
    quotaRemaining: s.quotaRemaining,
    subscriptionStatus: s.subscriptionStatus,
    subscriptionActive: s.subscriptionActive,
    billingModelType: s.billingModelType,
    measurementType: s.measurementType,
    unitLabel: s.unitLabel,
    signingVersion: request.signingVersion,
  });
}

function parseSignedUserContextFromHeaders(headers: HeaderBag): SignedUserContext {
  const rolesHeader = headerGet(headers, TollaraHeaders.ROLES);
  const roles = rolesHeader ? rolesHeader.split(',').map((x) => x.trim()).filter(Boolean) : [];
  let quotaRemaining: number | string | null = headerGet(headers, TollaraHeaders.QUOTA_REMAINING);
  if (quotaRemaining != null && quotaRemaining !== '') {
    const n = Number(quotaRemaining);
    if (!Number.isNaN(n)) quotaRemaining = n;
  } else {
    quotaRemaining = null;
  }
  const subRaw = headerGet(headers, TollaraHeaders.SUBSCRIPTION_ACTIVE);
  const subscriptionActive = parseSubscriptionActive(subRaw);
  const billing = headerGet(headers, TollaraHeaders.BILLING_MODEL);
  const measurement = headerGet(headers, TollaraHeaders.MEASUREMENT_TYPE);
  const unit = headerGet(headers, TollaraHeaders.UNIT_LABEL);
  return {
    userId: headerGet(headers, TollaraHeaders.USER_ID),
    serviceProductId: headerGet(headers, TollaraHeaders.SERVICE_PRODUCT_ID),
    plan: headerGet(headers, TollaraHeaders.PLAN),
    roles,
    quotaRemaining,
    subscriptionStatus: headerGet(headers, TollaraHeaders.SUBSCRIPTION_STATUS),
    subscriptionActive,
    billingModelType: billing && billing !== '' ? billing : null,
    measurementType: measurement && measurement !== '' ? measurement : null,
    unitLabel: unit && unit !== '' ? unit : null,
  };
}

/**
 * Verifies inbound HMAC from a case-insensitive header bag and payload.
 */
export function verifySignatureFromHeaders(
  serviceSecret: string,
  headers: HeaderBag,
  payload: string | object | null
): boolean {
  const signature = headerGet(headers, TollaraHeaders.SIGNATURE);
  const timestamp = headerGet(headers, TollaraHeaders.TIMESTAMP);
  if (!signature || !timestamp) return false;
  const signedUserContext = parseSignedUserContextFromHeaders(headers);
  const signingVersion = headerGet(headers, TollaraHeaders.SIGNING_VERSION);
  return verifyInboundHmac(serviceSecret, {
    signature,
    timestamp,
    payload,
    signedUserContext,
    signingVersion: signingVersion && signingVersion !== '' ? signingVersion : null,
  });
}

/**
 * Verifies inbound HMAC; if valid returns user context, else `null` (do not trust headers).
 */
export function verifySignatureFromHeadersAndGetUserContext(
  serviceSecret: string,
  headers: HeaderBag,
  payload: string | object | null
): UserContext | null {
  if (!verifySignatureFromHeaders(serviceSecret, headers, payload)) return null;
  return getUserContext(headers);
}

export function getUserContext(headers: HeaderBag): UserContext {
  const s = parseSignedUserContextFromHeaders(headers);
  return {
    userId: s.userId,
    serviceProductId: s.serviceProductId ?? null,
    plan: s.plan ?? null,
    roles: s.roles ?? [],
    quotaRemaining:
      typeof s.quotaRemaining === 'number'
        ? s.quotaRemaining
        : s.quotaRemaining != null && s.quotaRemaining !== ''
          ? Number(s.quotaRemaining)
          : null,
    subscriptionStatus: s.subscriptionStatus ?? null,
    subscriptionActive: s.subscriptionActive ?? false,
    billingModelType: s.billingModelType ?? null,
    measurementType: s.measurementType ?? null,
    unitLabel: s.unitLabel ?? null,
  };
}

File: sdk-python/src/agentvend_sdk/verifier.py
import json
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union

from .agentvend_headers import AgentVendHeaders
from .hmac_utils import calculate_hmac, constant_time_equals


@dataclass
class UserContext:
    user_id: Optional[str]
    plan: Optional[str]
    roles: List[str]
    quota_remaining: Optional[Union[int, float]]
    subscription_active: bool
    billing_model_type: Optional[str]
    measurement_type: Optional[str]
    unit_label: Optional[str]


@dataclass
class SignedUserContext:
    """Fields that participate in inbound HMAC userContextString (docs/hmac-spec.md)."""

    user_id: Optional[str] = None
    plan: Optional[str] = None
    roles: List[str] = field(default_factory=list)
    quota_remaining: Optional[Union[int, float]] = None
    subscription_active: bool = False
    billing_model_type: Optional[str] = None
    measurement_type: Optional[str] = None
    unit_label: Optional[str] = None


@dataclass
class InboundHmacRequest:
    signature: str
    timestamp: str
    payload: Any
    signed_user_context: SignedUserContext


def _header_get_ci(headers: Dict[str, Optional[str]], canonical_name: str) -> Optional[str]:
    target = canonical_name.lower()
    for k, v in headers.items():
        if k is not None and k.lower() == target:
            return v
    return None


def _format_quota(quota_remaining: Optional[Union[int, float, str]]) -> str:
    if quota_remaining is None:
        return ""
    if isinstance(quota_remaining, str):
        return quota_remaining
    if isinstance(quota_remaining, int):
        return str(quota_remaining)
    if quota_remaining == int(quota_remaining):
        return str(int(quota_remaining))
    return str(quota_remaining)


def _parse_quota_raw(raw: Optional[str]) -> Optional[Union[int, float]]:
    if raw in (None, ""):
        return None
    s = raw.strip()
    if "." in s:
        try:
            return float(s)
        except ValueError:
            return None
    try:
        return int(s)
    except ValueError:
        try:
            return float(s)
        except ValueError:
            return None


def _parse_subscription_active(raw: Optional[str]) -> bool:
    if raw is None or raw == "":
        return False
    t = raw.strip()
    return t.lower() == "true" or t == "1"


def build_gateway_user_context_string(
    user_id: Optional[str],
    plan: Optional[str],
    roles: List[str],
    quota_remaining: Optional[Union[int, float, str]],
    subscription_active: bool,
    billing_model_type: Optional[str],
    measurement_type: Optional[str],
    unit_label: Optional[str],
) -> str:
    u = user_id or ""
    p = plan or ""
    r = ",".join(roles) if roles else ""
    q = _format_quota(quota_remaining)
    sub = "true" if subscription_active else "false"
    b = billing_model_type or ""
    m = measurement_type or ""
    ul = unit_label or ""
    return u + p + r + q + sub + b + m + ul


def verify_inbound_hmac(agent_secret: str, request: InboundHmacRequest) -> bool:
    s = request.signed_user_context
    return verify_signature(
        agent_secret,
        request.signature,
        request.timestamp,
        request.payload,
        s.user_id,
        s.plan,
        s.roles,
        s.quota_remaining,
        s.subscription_active,
        s.billing_model_type,
        s.measurement_type,
        s.unit_label,
    )


def verify_signature_from_headers(
    agent_secret: str,
    headers: Dict[str, Optional[str]],
    payload: Any,
) -> bool:
    signature = _header_get_ci(headers, AgentVendHeaders.SIGNATURE) or ""
    timestamp = _header_get_ci(headers, AgentVendHeaders.TIMESTAMP) or ""
    if not signature or not timestamp:
        return False
    roles_str = _header_get_ci(headers, AgentVendHeaders.ROLES) or ""
    roles = [s.strip() for s in roles_str.split(",") if s.strip()]
    q_raw = _header_get_ci(headers, AgentVendHeaders.QUOTA_REMAINING)
    quota_remaining = _parse_quota_raw(q_raw)
    sub_raw = _header_get_ci(headers, AgentVendHeaders.SUBSCRIPTION_ACTIVE)
    bm = _header_get_ci(headers, AgentVendHeaders.BILLING_MODEL)
    mt = _header_get_ci(headers, AgentVendHeaders.MEASUREMENT_TYPE)
    ul = _header_get_ci(headers, AgentVendHeaders.UNIT_LABEL)
    signed = SignedUserContext(
        user_id=_header_get_ci(headers, AgentVendHeaders.USER_ID),
        plan=_header_get_ci(headers, AgentVendHeaders.PLAN),
        roles=roles,
        quota_remaining=quota_remaining,
        subscription_active=_parse_subscription_active(sub_raw),
        billing_model_type=bm if bm else None,
        measurement_type=mt if mt else None,
        unit_label=ul if ul else None,
    )
    return verify_inbound_hmac(
        agent_secret,
        InboundHmacRequest(
            signature=signature,
            timestamp=timestamp,
            payload=payload,
            signed_user_context=signed,
        ),
    )


def verify_signature(
    agent_secret: str,
    signature: str,
    timestamp: str,
    payload: Any,
    user_id: Optional[str],
    plan: Optional[str],
    roles: List[str],
    quota_remaining: Optional[Union[int, float, str]],
    subscription_active: bool,
    billing_model_type: Optional[str] = None,
    measurement_type: Optional[str] = None,
    unit_label: Optional[str] = None,
) -> bool:
    if not signature or not timestamp or not agent_secret:
        return False
    try:
        payload_string = "" if payload is None else (payload if isinstance(payload, str) else json.dumps(payload))
        user_context_string = build_gateway_user_context_string(
            user_id,
            plan,
            roles,
            quota_remaining,
            subscription_active,
            billing_model_type,
            measurement_type,
            unit_label,
        )
        data_to_sign = payload_string + timestamp + user_context_string
        expected = calculate_hmac(data_to_sign, agent_secret)
        return constant_time_equals(expected, signature)
    except Exception:
        return False


def verify_signature_from_headers_and_get_user_context(
    agent_secret: str,
    headers: Dict[str, Optional[str]],
    payload: Any,
) -> Optional[UserContext]:
    """Verify inbound HMAC; if valid return :class:`UserContext`, else ``None`` (do not trust headers)."""
    if not verify_signature_from_headers(agent_secret, headers, payload):
        return None
    return get_user_context(headers)


def get_user_context(headers: Dict[str, Optional[str]]) -> UserContext:
    roles_str = _header_get_ci(headers, AgentVendHeaders.ROLES) or ""
    roles = [s.strip() for s in roles_str.split(",") if s.strip()]
    q = _header_get_ci(headers, AgentVendHeaders.QUOTA_REMAINING)
    quota_remaining = _parse_quota_raw(q)
    sub = _header_get_ci(headers, AgentVendHeaders.SUBSCRIPTION_ACTIVE)
    subscription_active = _parse_subscription_active(sub)
    uid = _header_get_ci(headers, AgentVendHeaders.USER_ID)
    plan = _header_get_ci(headers, AgentVendHeaders.PLAN)
    bm = _header_get_ci(headers, AgentVendHeaders.BILLING_MODEL)
    mt = _header_get_ci(headers, AgentVendHeaders.MEASUREMENT_TYPE)
    ul = _header_get_ci(headers, AgentVendHeaders.UNIT_LABEL)
    return UserContext(
        user_id=uid,
        plan=plan,
        roles=roles,
        quota_remaining=quota_remaining,
        subscription_active=subscription_active,
        billing_model_type=bm if bm else None,
        measurement_type=mt if mt else None,
        unit_label=ul if ul else None,
    )
